File: 994.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<int> q_row, q_col; 
        vector<vector<int>> q ; 
        bool time_zero = true; 
        for (int i = 0 ; i < grid.size(); i++){
            for (int j = 0 ; j < grid[0].size(); j++){
                if (grid[i][j] == 2){
                    q_row.push(i); 
                    q_col.push(j); 
                    grid[i][j] = 3; 
     
                } 
                time_zero = grid[i][j] == 1? 0: time_zero; 
            }
        }
        
        if (time_zero) return 0; 
        
        int minutes = -1; 
        while(!q_row.empty()){
            queue<int> q_row_aux; 
            queue<int> q_col_aux; 
            while(!q_row.empty()){
                int row = q_row.front(); 
                int col = q_col.front(); 
                q_row.pop();
                q_col.pop(); 

                if (validRange(grid, row+1, col)){
                    q_row_aux.push(row+1); 
                    q_col_aux.push(col); 
                    grid[row+1][col] = 3; 
                    
                }
                if (validRange(grid, row-1, col)){
                    q_row_aux.push(row-1);
                    q_col_aux.push(col); 
                    grid[row-1][col] = 3; 
                }
                if (validRange(grid, row, col+1)){
                    q_row_aux.push(row);
                    q_col_aux.push(col+1); 
                    grid[row][col+1] = 3; 
                }
                if (validRange(grid, row, col-1)){
                    q_row_aux.push(row);
                    q_col_aux.push(col-1); 
                    grid[row][col-1] = 3; 
                }
              
                 
            }
            minutes++; 
            q_row = q_row_aux; 
            q_col = q_col_aux; 
        }
        
        for (int i = 0 ; i < grid.size(); i++){
            for (int j = 0 ; j < grid[0].size(); j++){
                if (grid[i][j] == 1) return -1; 
            }
        }
        return minutes; 
        
        
    }
    
    int validRange(vector<vector<int>>& grid, int row,int col){
        return row >= 0 && row < grid.size() && col >= 0 && col < grid[0].size() && grid[row][col] == 1;  
    }

    
};

File: 994.py
import queue
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = queue.Queue()
        m = len(grid)
        n = len(grid[0]) 
        notRotting = 0

        # Find routing and adds to the queue
        for i, line in enumerate(grid):
            for j, orange in enumerate(line): 
                if orange == 2: q.put([i,j])
                elif orange == 1: notRotting += 1
        
        # No rotting orange. 
        if q.empty() and notRotting != 0:
            return -1
        # We don't have oranges.
        elif notRotting == 0:
            return 0
        

        def getAdjacentRotting(i: int, j: int):
            if i > 0 and grid[i-1][j] == 1: yield [i-1,j]
            if j > 0 and grid[i][j-1] == 1: yield [i,j-1]
            if i < m-1 and grid[i+1][j] == 1: yield [i+1,j]
            if j < n-1 and grid[i][j+1] == 1: yield [i,j+1] 

        days = 0
        while not q.empty():
            temp = []   
            while not q.empty():
                orange = q.get()
                for newRotting in getAdjacentRotting(orange[0], orange[1]): 
                    temp.append(newRotting)
                    grid[newRotting[0]][newRotting[1]] = 2
                    notRotting-= 1
                    
            for newRotting in temp:
                q.put(newRotting)
            
            days += 1
        
        # Orange cannot be reached. 
        if notRotting > 0: 
            return -1
        
        # Have answer.
        return days -1 







